### DISCUSSION TITLE
How to make chart template for use with different data sources

### DISCUSSION THREAD

**Discussion by mcp292** [vega/altair#2826] [Q&A] [answered]
# MWE
```py
import altair as alt
import numpy as np
import pandas as pd

x, y = np.meshgrid(range(-5, 5), range(-5, 5))
z1 = x ** 2 + y ** 2
z2 = x * 2 + y * 2

df1 = pd.DataFrame({'x': x.ravel(),
		    'y': y.ravel(),
		    'z': z1.ravel()})

df2 = pd.DataFrame({'x': x.ravel(),
		    'y': y.ravel(),
		    'z': z2.ravel()})

template = alt.Chart().mark_rect().encode(
    x='x:O',
    y='y:O',
    color='z:Q'
)

df1_chart = template.data(df1)
df2_chart = template.data(df2)
```

Modifed from [this example](https://altair-viz.github.io/gallery/simple_heatmap.html).

# Question
I'm looking for something semantically equivalent to the last two lines, but syntactically correct.

I am looking for an altair way to do this; I understand I could wrap the chart code in a function.

**Comment by joelostblom** [comment, marked as answer]
You can either use the properties method:

```py
template.properties(data=df1)
```

Or directly modify the object

```py
template['data'] = df1
template
```

The first would likely be to prefer in most cases since it uses a method rather than changing the object manually

**Reply by mcp292** [reply]
Outstanding!! Thank you!
